chronos/futures.nim

```diff
@@ -0,0 +1,369 @@
+          if entry.line == reraisedFromBegin:
+            result.add(spaces(indent) & "#[\n")
+            indent.inc(2)
+          elif entry.line == reraisedFromEnd:
+            indent.dec(2)
+            result.add(spaces(indent) & "]#\n")
+          continue
+
+        let left = "$#($#)" % [filename, $entry.line]
+        result.add((spaces(indent) & "$#$# $#\n") % [
+          left,
+          spaces(longestLeft - left.len + 2),
+          procname
+        ])
+        let hint = getHint(entry)
+        if hint.len > 0:
+          result.add(spaces(indent+2) & "## " & hint & "\n")
+    except ValueError as exc:
+      return exc.msg # Shouldn't actually happen since we set the formatting
+                    # string
+
+  proc injectStacktrace(future: FutureBase) =
+    const header = "\nAsync traceback:\n"
+
+    var exceptionMsg = future.error.msg
+    if header in exceptionMsg:
+      # This is messy: extract the original exception message from the msg
+      # containing the async traceback.
+      let start = exceptionMsg.find(header)
+      exceptionMsg = exceptionMsg[0..<start]
+
+    var newMsg = exceptionMsg & header
+
+    let entries = getStackTraceEntries(future.error)
+    newMsg.add($entries)
+
+    newMsg.add("Exception message: " & exceptionMsg & "\n")
+
+    # # For debugging purposes
+    # newMsg.add("Exception type:")
+    # for entry in getStackTraceEntries(future.error):
+    #   newMsg.add "\n" & $entry
+    future.error.msg = newMsg
+
+# Internal utilities - these are not part of the stable API
+proc internalCheckComplete*(fut: FutureBase) {.
+     raises: [Defect, CatchableError].} =
+  # For internal use only. Used in asyncmacro
+  if not(isNil(fut.error)):
+    when chronosStackTrace:
+      injectStacktrace(fut)
+    raise fut.error
+
+template internalLocation*(fut: FutureBase): array[LocationKind, ptr SrcLoc] = fut.location
+template internalCallbacks*(fut: FutureBase): seq[AsyncCallback] = fut.callbacks
+template internalState*(fut: FutureBase): FutureState = fut.state
+template internalChild*(fut: FutureBase): FutureBase = fut.child
+template internalCancelcb*(fut: FutureBase): CallbackFunc = fut.cancelcb
+template internalMustCancel*(fut: FutureBase): bool = fut.mustCancel
+template internalError*(fut: FutureBase): ref CatchableError = fut.error
+template internalValue*[T](fut: Future[T]): T = fut.value
+
+func location*(fut: FutureBase): array[LocationKind, ptr SrcLoc] = fut.location
+func state*(fut: FutureBase): FutureState = fut.state
+
+when defined(chronosV4):
+  # `error` will change definition when we move to V4
+  func error*(fut: FutureBase): ref CatchableError {.
+      deprecated: "Use `readError` to access error of future".} = fut.error
+
+  func `[]`*(loc: array[LocationKind, ptr SrcLoc], v: int): ptr SrcLoc =
+    case v
+    of 0: loc[LocationKind.Create]
+    of 1: loc[LocationKind.Complete]
+    else: raiseAssert("Unknown source location " & $v)
+
+proc read*[T](future: Future[T] ): T {.
```

Comment:
This procedure should also be `template`.

Reply:
hm, a template would have to protect against multiple evaluations of `future` - I think the difference is minimal - once we fully move to 1.6, we can return `lent` instead which will remove any practical performance difference vs a template, wdyt?

Reply:
Ditto.

Reply:
`await` uses `internalValue` directly, without calling `read`